### PR TITLE
github-actions: enable provenance for the jar files

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

Enable the provenance for the jar files generated in Buildkite using the [GitHub provenance action](https://github.com/github-early-access/generate-build-provenance).

## Test
I tested it in [here](https://github.com/elastic/apm-agent-java/actions/runs/8819973880) and it produced these attestations:
- https://github.com/elastic/apm-agent-java/attestations

## Checklist

- [ ] This is an enhancement of existing features, or a new feature in existing plugins
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] I have made corresponding changes to the documentation
- [ ] This is a bugfix
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] I have added tests that would fail without this fix
- [ ] This is a new plugin
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] My code follows the [style guidelines of this project](https://github.com/elastic/apm-agent-java/blob/main/CONTRIBUTING.md#java-language-formatting-guidelines)
  - [ ] I have made corresponding changes to the documentation
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/main/CONTRIBUTING.md#testing) pass locally with my changes
  - [ ] I have updated [supported-technologies.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/docs/supported-technologies.asciidoc)
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] Added an instrumentation plugin? Describe how you made sure that old, non-supported versions are not instrumented by accident.
- [ ] This is something else
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)

### CODE REVIEW

**Review by reakaleek — APPROVED** [review]

File: .ci/release.sh [hunk 1]
```diff
@@ -32,3 +32,6 @@ fi
 
 echo "--- Deploy the release :package: [./mvnw $GOAL)] $DRY_RUN_MSG"
 ./mvnw -V -s .ci/settings.xml -Pgpg clean $GOAL -DskipTests --batch-mode | tee release.txt
+
+echo "--- Archive the target folder with jar files"
+find . -type d -name target -exec find {} -name '*.jar' -print0 \; | xargs -0 tar -cvf "${TARBALL_FILE:-dist.tar}"
```
Comment (SylvainJuge) [inline .ci/release.sh:37]: Use `.ci/scripts/published-artifacts-list.sh` script added in #3595 instead to get the list of published artifacts

File: .ci/snapshot.sh [hunk 2]
```diff
@@ -34,3 +34,6 @@ fi
 
 echo "--- Deploy the snapshot :package: [./mvnw $GOAL)] $DRY_RUN_MSG"
 ./mvnw -V -s .ci/settings.xml -Pgpg clean $GOAL -DskipTests --batch-mode | tee snapshot.txt
+
+echo "--- Archive the target folder with jar files"
+find . -type d -name target -exec find {} -name '*.jar' -print0 \; | xargs -0 tar -cvf "${TARBALL_FILE:-dist.tar}"
```
Comment (SylvainJuge) [inline .ci/snapshot.sh:39]: Use `.ci/scripts/published-artifacts-list.sh` script added in #3595 instead to get the list of published artifacts

**Review by SylvainJuge — CHANGES_REQUESTED** [review]

File: .ci/release.sh (hunk 1 above)
Comment (v1v) [reply .ci/release.sh]: see https://github.com/elastic/apm-agent-java/pull/3594/commits/ce87909b089a09acaa879bfc8fe73eae3ad890ca

File: .ci/snapshot.sh (hunk 2 above)
Comment (v1v) [reply .ci/snapshot.sh]: see https://github.com/elastic/apm-agent-java/pull/3594/commits/ce87909b089a09acaa879bfc8fe73eae3ad890ca

**Review by SylvainJuge — APPROVED** [review]

**Pull request merged**